### PR TITLE
Ps 5.7 myr 147

### PR DESCRIPTION
**Pull request by george-lorch**
Jenkins just started. Prior and local runs look good, lets see what this turns out http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1121/

### CODE REVIEW

**Comment by george-lorch** [comment]
Bad job, fat finger on mtr args, try http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1122/

File: mysql-test/suite/rocksdb/t/ttl_primary.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/ttl_primary.test:175]: Unstable test (requires < 5s wall time since INSERT, which is not guaranteed). I am not sure whether it applies to this test directly, but upstream have switched to using debug variables for determinism in this scenarios, perhaps it's in the upstream merge queue.

File: mysql-test/suite/rocksdb/t/ttl_primary.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/ttl_primary.test:254]: Likewise

File: mysql-test/suite/rocksdb/t/ttl_primary.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/ttl_primary.test:386]: Likewise

File: mysql-test/suite/rocksdb/t/ttl_primary.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/ttl_primary.test:413]: Likewise

File: mysql-test/suite/rocksdb/t/ttl_primary.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/ttl_primary.test:438]: Likewise

File: mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test:1]: --source include/have_partition.inc

File: mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test:82]: Likewise

File: mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test:140]: Likewise

File: mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test:189]: Likewise

File: mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test:233]: Likewise

File: mysql-test/suite/rocksdb/t/use_direct_reads_writes.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: Not in the scope of this PR, but this whole testcase must be cleaned up WRT error log handling

File: storage/rocksdb/ha_rocksdb.cc
Comment (laurynas-biveinis) [inline storage/rocksdb/ha_rocksdb.cc]: TRUE

File: storage/rocksdb/ha_rocksdb.cc
Comment (laurynas-biveinis) [inline storage/rocksdb/ha_rocksdb.cc]: FALSE

File: storage/rocksdb/ha_rocksdb.cc
Comment (laurynas-biveinis) [inline storage/rocksdb/ha_rocksdb.cc]: Two different styles of "unused" (I'd prefer the former as shorter)

File: storage/rocksdb/rdb_datadic.h
Comment (laurynas-biveinis) [inline storage/rocksdb/rdb_datadic.h]: constexpr (no need to allocate storage)

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: mysql-test/suite/rocksdb/t/ttl_primary.test
Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/ttl_primary.test:175]: True. There are two commits coming from upstream, one in next batch (May) and one in the following (June) that work towards addressing flakyness:
https://github.com/facebook/mysql-5.6/commit/ad90bf77f3a353336c0e33f5784eb9b30039880a
https://github.com/facebook/mysql-5.6/commit/891a9115e3d16e1a2b2efab11eff1880eb69ce1e

I would prefer to keep these aligned with upstream for now until I get through the merges and if we still have stability issues, then we can work through fixing them and sending the changes upstream.

File: mysql-test/suite/rocksdb/t/use_direct_reads_writes.test
Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: I had cleaned up some stuff already in this series and missed this one, I'll fix it.

File: storage/rocksdb/ha_rocksdb.cc
Comment (george-lorch) [reply storage/rocksdb/ha_rocksdb.cc]: Actually, these are all GLOBAL sys_vars, they should not need initialization at all as the defaults should be set via the MYSQL_SYSVAR_* macros and init process, no?

Comment (laurynas-biveinis) [reply storage/rocksdb/ha_rocksdb.cc]: No! Unless MyRocks has explicit code to iterate over its sysvars, they are not set up during init process. You'd think it's a natural thing to have, but I've been burnt several times by this and have seen others trip over the same assumption

Comment (laurynas-biveinis) [reply storage/rocksdb/ha_rocksdb.cc]: Actually the SE must have some startup init code, or setting the vars through command line would simply be ignored

Comment (george-lorch) [reply storage/rocksdb/ha_rocksdb.cc]: Really? Interesting, then what is the point of a 'default' value for the GLOBAL SYSVAR macros. Seems to me it should be one or the other, explicit init of the static or explicit init via SYSVAR default, not some weird mix of both. Either way, I'll fix it but they have many my_bools just using 0|1 so I think rather than just fixing these in their cherry pick commits, I will leave them as merged and append a commit of my own to the whole change set that fixed them all in one shot and watch for more of the same in future merges.

Comment (george-lorch) [reply storage/rocksdb/ha_rocksdb.cc]: This is going to be nice, MyRocks uses a static global singleton `std::unique_ptr<rocksdb::DBOptions> rocksdb_db_options = rdb_init_rocksdb_db_options();` to get default values for the SYSVARS macros and thus eventually the sysvar values, following on the belief/assumption I made above. So if what you say is true, then they have potential startup init and use ordering issues waiting to happen.

Comment (george-lorch) [reply storage/rocksdb/ha_rocksdb.cc]: And we have a winner:
`static unsigned long rocksdb_persistent_cache_size_mb;`
...
`static MYSQL_SYSVAR_ULONG(
    persistent_cache_size_mb, rocksdb_persistent_cache_size_mb,
    PLUGIN_VAR_RQCMDARG | PLUGIN_VAR_READONLY,
    "Size of cache in MB for BlockBasedTableOptions::persistent_cache "
    "for RocksDB", nullptr, nullptr, rocksdb_persistent_cache_size_mb,
    /* min */ 0L, /* max */ ULONG_MAX, 0);
`
<<< Using uninitialized static for the default value in the macro. This value is not explicitly set anywhere in the code. This may explain some of the instability I have with the persistent_cache test.

Comment (george-lorch) [reply storage/rocksdb/ha_rocksdb.cc]: Sure enough, my persistent_cache test instability is gone by fixing up the variable inits. This is a very good and worthwhile effort.

**Comment by george-lorch** [comment]
OK, addressed the agreed stuff and also fixed another missed issue in one of the commits for rocksdb.type_varchar that was missing a --loose- prefix in .opt file.
On top of that I added 3 new and very minor commits for:
https://jira.percona.com/browse/MYR-163 : global sysvars initialized incorrectly
https://jira.percona.com/browse/MYR-164 : rocksdb.persistent_cache needs cleanup and re-record to avoid fouling mtr server err log
https://jira.percona.com/browse/MYR-165 : rocksdb.collation does not clean up alternate err log from tmp

Respun jenkins on all of this here : http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1125/

File: mysql-test/suite/rocksdb/t/persistent_cache.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/persistent_cache.test]: Based on the comment next to the added exec rm -rf at the bottom, this is not a "file" but "directory"?

File: mysql-test/suite/rocksdb/t/persistent_cache.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/persistent_cache.test]: I'd mildly prefer masking out only the non-constant $MYSQLTEST_VARDIR part (IIRC I have suggested this before but you haven't stated your preference so I can keep or stop suggesting this)

File: mysql-test/suite/rocksdb/t/persistent_cache.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/persistent_cache.test]: Likewise

File: mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test
Comment (laurynas-biveinis) [reply mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test:1]: Still --source include/have_partition.inc unless, as with the previous test, there is something in the next merge?

File: mysql-test/suite/rocksdb/t/use_direct_reads_writes.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: I'd swap these two lines (now it looks like the crash is expected on shutdown)

Comment (laurynas-biveinis) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: And replace shutdown_server with shutdown_mysqld.inc. That it will write a non-crash-expecting MTR file is immaterial, it will be overwritten by expect_crash.inc, OTOH the "10" timeout is very problematic with Valgrind runs etc.

File: mysql-test/suite/rocksdb/t/use_direct_reads_writes.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: This works, but with an extremely long timeout (even worse with Valgrind). The canonical MTR way is --error, --exec $MYSQLD (grep main testsuite for examples)

File: mysql-test/suite/rocksdb/t/use_direct_reads_writes.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: Likewise

File: mysql-test/suite/rocksdb/t/use_direct_reads_writes.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: Likewise

File: storage/rocksdb/ha_rocksdb.cc
Comment (laurynas-biveinis) [inline storage/rocksdb/ha_rocksdb.cc]: constexpr?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
The variable initialisation bugfix is better upstreamed sooner than later, right?

File: mysql-test/suite/rocksdb/t/persistent_cache.test
Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/persistent_cache.test]: This does not work as you suggest, tried it, seems env var is not picked up through the process. Maybe you have an exact similar example where this does work with $restart_parameters.

Comment (laurynas-biveinis) [reply mysql-test/suite/rocksdb/t/persistent_cache.test]: mysql-test/include/percona_signal_handling.inc:

```
--let $test_error_log= $MYSQLTEST_VARDIR/tmp/test_threadpool_sig_handler.err
--let $restart_parameters=restart:--log-error=$test_error_log
--replace_result $MYSQLTEST_VARDIR MYSQLTEST_VARDIR
--source include/restart_mysqld.inc

```

Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/persistent_cache.test]: Ahh, I totally misunderstood your initial request/idea and get it now.

File: mysql-test/suite/rocksdb/t/persistent_cache.test
Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/persistent_cache.test]: Yeah, correct. This is/was upstream naming convention and I really didn't know until I dug into the test but didn't go back and rename the vars.

File: mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test
Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/ttl_primary_with_partitions.test:1]: Doesn't exist in upstream, I'll just add it to ours on the commit introducing the test and put it into the great list of upstream mtr diffs.

File: mysql-test/suite/rocksdb/t/use_direct_reads_writes.test
Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: There is no _exact_ example that matches this case with a plugin load needed. For some reason the changes I pushed run correctly locally, but fail across 100% of our jenkins farm:
> 17:26:54 rocksdb.use_direct_reads_writes          w3 [ fail ]
> 17:26:54         Test ended at 2017-08-15 21:26:54
> 17:26:54 
> 17:26:54 CURRENT_TEST: rocksdb.use_direct_reads_writes
> 17:26:54 mysqld-debug: Can't change dir to '/var/lib/mysql/' (Errcode: 2 - No such file or directory)
> 17:26:54 2017-08-15T21:26:50.065769Z 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server option (see documentation for more details).
> 17:26:54 2017-08-15T21:26:50.065865Z 0 [Note] --secure-file-priv is set to NULL. Operations related to importing and exporting data are disabled
> 17:26:54 2017-08-15T21:26:50.065897Z 0 [Note] /mnt/workspace/mysql-5.7-param/BUILD_TYPE/debug/Host/ubuntu-trusty-64bit/build/sql/mysqld-debug (mysqld 5.7.18-16-debug) starting as process 17851 ...
> 17:26:54 2017-08-15T21:26:50.065941Z 0 [ERROR] Can't find error-message file '/usr/share/mysql/errmsg.sys'. Check error-message file location and 'lc-messages-dir' configuration directive.
> 17:26:54 2017-08-15T21:26:50.067389Z 0 [Warning] Can't create test file /var/lib/mysql/os-trusty-x64-2030.lower-test
> 17:26:54 2017-08-15T21:26:50.067408Z 0 [Warning] Can't create test file /var/lib/mysql/os-trusty-x64-2030.lower-test
> 17:26:54 2017-08-15T21:26:50.067811Z 0 [ERROR] failed to set datadir to /var/lib/mysql/
> 17:26:54 2017-08-15T21:26:50.067823Z 0 [ERROR] Aborting

I suspect this might be some odd difference between in-source and out-of-source builds. 
I have tried other ways of invoking --exec $MYSQLD based on numerous existing tests but every combination seems to run into one issue or another. So I'm about out of ideas here how this is supposed to work and can't find anything reasonably close in any documentation or existing suites.

Please take a look at the pushed changes and see if there is something obvious that I am missing here.

Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: Actually, I think I might have just stumbled on the correct incantation...stay tuned...

Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: Nope, but I have confirmed that this behaves very differently between an in-source and out-of-source build. out-of-source works fine, in-source doesn't.

Comment (laurynas-biveinis) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: @georgelorchpercona what is the incantation you tested that works for out-of-source build for you is the one pushed right now, right?
```
--error 1
--exec $MYSQLD --log-error=$SEARCH_FILE --rocksdb_use_direct_reads=1 --rocksdb_allow_mmap_reads=1
```

Comment (laurynas-biveinis) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: @georgelorchpercona , try $MYSQLD_CMD instead of $MYSQLD, this is the MTR difference between the two:

```
  my $exe_mysqld= find_mysqld($basedir);
  $ENV{'MYSQLD'}= $exe_mysqld;
  my $extra_opts= join (" ", @opt_extra_mysqld_opt);
  $ENV{'MYSQLD_CMD'}= "$exe_mysqld --defaults-group-suffix=.1 ".
    "--defaults-file=$path_config_file $extra_opts";

```

Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: Correct, actually I think it is not just out-of-source but using INSTALL_PREFIX as well.

Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: Yup, tried that. Without overriding all of the defaults with --no-defaults, the problem is that the rocksdb plugin is not loaded. Adding the $ROCKSDB_OPT and $ROCKSDB_LOAD or $ROCKSDB_LOAD_ADD just causes the server launch to hang. I believe it is because these options are not properly quoted for use on the command line via direct --exec

Comment (laurynas-biveinis) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: Can you post the resulting failing command line w/o $ROCKSDB_* options and the contents of $path_config_file?

Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: Yeah, this is bizzarre...it seems that trying to use '--exec $MYSQLD' and '--exec $MYSQLD_CMD' with any kind of plugin load just hangs, no output log files, nothing. There must be some issue with the way the new shell is being created. I even tried something as simple as --plugin_load=$ROCKSDB_DIR/$ROCKSDB --plugin_load_add=$ROCKSDB_DIR/$ROCKSDB and it just doesn't want to launch the process. Echoing out the -exec just shows

> /ssd/percona/myrocks/percona-server-5.7/sql/mysqld --defaults-group-suffix=.1 --defaults-file=/ssd/percona/myrocks/percona-server-5.7/mysql-test/var/my.cnf --loose-tokudb-cache-size=512M --log-syslog=0 --log-error=/ssd/percona/myrocks/percona-server-5.7/mysql-test/var/tmp/rocksdb.use_direct_reads_writes.err --plugin_load_add=/ssd/percona/myrocks/percona-server-5.7/storage/rocksdb/ha_rocksdb.so --rocksdb_use_direct_reads=1 --rocksdb_allow_mmap_reads=1

Which looks fine to me :-\



This has got to be something totally stupid and obvious that I am missing.

Comment (laurynas-biveinis) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: What about defaults-file=...my.cnf contents? What does machine ps show during the hung launch? Does MTR --debug or --verbose (both can be repeated several times IIRC) show what it's actually trying to invoke or what it is waiting for? We can take this offline from PR review if more convenient

Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: w/o the ROCKSDB_* options, the --exec succeeds, but then the test fails because the options being tested for mutual exclusive values don't exist as the plugin is not loaded.

Yes, lets just drop to email...

Comment (laurynas-biveinis) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test]: Does this refer to ROCKSDB_ option added manually for testing or are they present in the $MYSQLD_CMD? If it's not the latter, then regular testcase restarts shouldn't work neither, suspicious

**Comment by george-lorch** [comment]
As discussed in slack, root of test problem found and test is now clean and clear. Jenkins here http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1150/

**Comment by george-lorch** [comment]
Worth noting, as mentioned above, the unstable ttl_* tests due to relying on sleeps for synchronization.

File: mysql-test/suite/rocksdb/t/use_direct_reads_writes.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/use_direct_reads_writes.test:6]: Great catch to fix the comment here ;)

File: mysql-test/suite/rocksdb/t/use_direct_reads_writes.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rocksdb/t/use_direct_reads_writes.test:25]: Is there a chance of SEARCH_PATTERN mix-up, i.e. the 2nd restart picking up something printed by the 1st one? If none at all, OK to keep as-is, otherwise the 2nd restart should use a different error log

**Review by laurynas-biveinis — APPROVED** [review]

File: mysql-test/suite/rocksdb/t/use_direct_reads_writes.test
Comment (george-lorch) [reply mysql-test/suite/rocksdb/t/use_direct_reads_writes.test:25]: No, the search text was chosen specifically to avoid that.

**Pull request merged**
